executor/code_loader.py

```diff
@@ -0,0 +1,112 @@
+
+    # Predefined timeouts to be used:
+    setup_timeout = 2.0
+    interval = 0.01
+    TIMEOUT = {
+        'autonomous': 5.0,
+        'teleop': 120.0, 
+    }
+    
+
+    def __init__(self, student_module: str, mode: str):
+        """
+        Initializes class attributes and loads the student code.
+        Args:
+            student_module: name of python file with student code, without the .py
+            mode: either 'autonomous' or 'teleop'
+        """
+        self.student_module = student_module
+        self.mode = mode
+        self.load_code()
+
+
+    def run_with_timeout(self, func: typing.Callable, timeout: float, *args, **kwargs):
+        """Runs any function `func` for the specified amount of time `timeout`. """
+        signal.setitimer(signal.ITIMER_REAL, timeout)
+        try:
+            return func(*args, **kwargs)
+        finally:
+            signal.setitimer(signal.ITIMER_REAL, 0)
+
+
+    def load_code(self):
+        """Loads student code into this file's namespace."""
+        self.student_code = importlib.import_module(self.student_module)
+        setattr(self.student_code, 'print', _print)
+        setattr(self.student_code, 'Robot', studentapi.Robot())
+        setattr(self.student_code, 'Gamepad', studentapi.Gamepad(self.mode))
+
+
+    def get_function_safe(self, name: str):
+        """Tries to get a function with name `name` from the student's module. If not found, returns a stub function that does nothing. """
+        try:
+            func = getattr(self.student_code, name)
+        except AttributeError:
+            _print(f'Unable to find function {name}, using stub')
+            def func(): return
+        return func
+
+
+    def run(self):
+        """Runs the student code with the predefined timeouts. First calls `mode_setup()` then calls `mode_main()` based on the `mode` attribute. The timeout is handled using SIGALRM signals."""
+        if self.mode != 'autonomous' and self.mode != 'teleop':
+            raise CodeExecutionError("Only modes that can be run are autonomous and teleop.")
+        setup = self.get_function_safe(f'{self.mode}_setup')
+        main = self.get_function_safe(f'{self.mode}_main')
+        start = time.time()
+        try:
+            self.run_with_timeout(setup, self.setup_timeout)
+        except Exception as e:
+            signal.setitimer(signal.ITIMER_REAL, 0)
+            raise CodeExecutionError(f"Error in {self.mode} setup: {repr(e)}")
+        diff = time.time() - start
+
+        signal.setitimer(signal.ITIMER_REAL, self.TIMEOUT[self.mode] - diff)
+        try:
+            while True:
+                start = time.time()
+                main()
+                diff = time.time() - start
+                if diff > self.interval:
+                    raise CodeExecutionError(f"An iteration of {self.mode}_main took {diff} seconds, longer than {self.interval} seconds, indicating student code is stuck in a loop.")
```

Comment:
Runtime should continue running student code to the best of its ability and log the poor performance instead of exiting early. This may cause robots to stop during a match due to the idiosyncrasies of the CPU scheduler.